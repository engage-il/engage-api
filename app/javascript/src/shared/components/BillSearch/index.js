// @flow
export { BillSearch, initialVariables } from './BillSearch'
