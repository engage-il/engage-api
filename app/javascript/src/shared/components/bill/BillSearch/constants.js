// @flow
export const BILL_SPACING = 40
export const BILL_SPACING_MOBILE = 30
