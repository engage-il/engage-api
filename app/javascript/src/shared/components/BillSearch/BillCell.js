// @flow
import * as React from 'react'
import { createFragmentContainer, graphql } from 'react-relay'
import { css } from 'glamor'
import { formatDate } from 'shared/date'
import type { Bill } from 'shared/types'
import { Button } from 'shared/components'
import { stylesheet, colors, mixins } from 'shared/styles'

type Props = {
  bill: Bill,
  className?: string
}

let BillCell = function BillCell ({
  bill
}: Props) {
  const formattedDate = formatDate(bill.hearing.date, 'DD/MM/YYYY')

  return (
    <div {...css(rules.container)}>
      <div {...rules.info}>
        <div {...rules.header}>
          <div {...rules.document}>
            <h3>{bill.title}</h3>
            <span>{bill.documentNumber}</span>
          </div>
          <p>{formattedDate}</p>
        </div>
        {bill.summary && <p {...rules.summary}>{bill.summary}</p>}
      </div>
      <div {...rules.actions}>
        <Button
          styles={rules.button}
          to={bill.witnessSlipUrl}
          children='Take Action'
        />
        <Button
          styles={rules.button}
          to={`/bill/${bill.id}`}
          isSecondary
          children='More Info'
        />
      </div>
    </div>
  )
}

BillCell = createFragmentContainer(BillCell, graphql`
  fragment BillCell_bill on Bill {
    id
    documentNumber
    title
    summary
    witnessSlipUrl
    detailsUrl
    fullTextUrl
    hearing {
      date
    }
  }
`)

const rules = stylesheet({
  container: {
    ...mixins.shadows.low,
    ...mixins.borders.low(),
    display: 'flex',
    padding: 15,
    marginBottom: 15,
    backgroundColor: colors.neutral,
    '&:last-child': {
      marginBottom: 0
    },
    ...mixins.mobile.glam({
      flexDirection: 'column'
    })
  },
  info: {
    ...mixins.borders.low(['right']),
    flex: 1,
    paddingRight: 15,
    ...mixins.mobile.glam({
      ...mixins.borders.low(['bottom']),
      paddingRight: 0,
      paddingBottom: 15,
      marginBottom: 15,
      borderRight: 'none'
    })
  },
  header: {
    display: 'flex',
    flexDirection: 'column'
  },
  document: {
    marginBottom: 5,
    '> *': {
      display: 'inline-block'
    },
    '> h3': {
      ...mixins.fonts.bold,
      marginRight: 10
    }
  },
  actions: {
    display: 'flex',
    flexDirection: 'column',
    paddingLeft: 15,
    ...mixins.mobile.glam({
      flexDirection: 'row',
      paddingLeft: 0
    })
  },
  button: {
    width: 200,
    marginBottom: 10,
    ':last-child': {
      marginBottom: 0
    },
    ...mixins.mobile.glam({
      flex: 1,
      width: 'auto',
      marginBottom: 0,
      marginRight: 10,
      ':last-child': {
        marginRight: 0
      }
    })
  },
  summary: {
    ...mixins.borders.low(['top']),
    marginTop: 10,
    paddingTop: 10
  }
})

export { BillCell }
