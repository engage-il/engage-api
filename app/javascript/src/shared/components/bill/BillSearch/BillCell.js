// @flow
import * as React from 'react'
import styled from 'react-emotion'
import { createFragmentContainer, graphql } from 'react-relay'
import { BillTitle } from '../BillTitle'
import { BillStatus } from '../BillStatus'
import { Button } from '@/components'
import { mixins } from '@/styles'
import type { Bill } from '@/types'

type Props = {
  bill: Bill,
  className?: string
}

let BillCell = function BillCell ({
  bill
}: Props) {
  return (
    <div>
      <Cell>
        <BillTitle bill={bill} />
        <BillStatus bill={bill} />
        <p>{bill.summary}</p>
        <Button to={`/bill/${bill.id}`} isSmall children='More Info' />
      </Cell>
    </div>
  )
}

BillCell = createFragmentContainer(BillCell, graphql`
  fragment BillCell_bill on Bill {
    id
    summary
    hearing {
      date
    }
    ...BillTitle_bill
    ...BillStatus_bill
  }
`)

const Cell = styled.div`
  ${mixins.flexColumn};

  align-items: flex-start;
  max-width: 960px;

  > p {
    margin-bottom: 15px;
  }
`

export { BillCell }
