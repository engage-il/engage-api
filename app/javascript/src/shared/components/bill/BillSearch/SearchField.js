// @flow
import * as React from 'react'
import styled, { css } from 'react-emotion'
import { Link } from '@/components'
import { colors, mixins } from '@/styles'
import type { SearchParams } from '@/types'

type Props = {
  params: SearchParams,
  onChange: (SearchParams) => void
}

type State = {
  isFocused: boolean
}

export class SearchField extends React.Component<Props, State> {
  state = {
    isFocused: false
  }

  // events
  didChangeQuery = ({ target }: SyntheticInputEvent<*>) => {
    const { params, onChange } = this.props

    onChange({
      ...params,
      query: target.value
    })
  }

  didChangeFocus = (isFocused: boolean) => () => {
    this.setState({ isFocused })
  }

  // lifecycle
  render () {
    const { params } = this.props
    const { isFocused } = this.state

    return (
      <Search>
        <FieldContainer>
          <Field isFocused={isFocused}>
            <Input
              type='text'
              name='search-field'
              value={params.query}
              placeholder={`health care, HB2364`}
              onChange={this.didChangeQuery}
              onFocus={this.didChangeFocus(true)}
              onBlur={this.didChangeFocus(false)}
            />
            <Button
              onClick={() => {}}
              children='Search'
            />
          </Field>
        </FieldContainer>
      </Search>
    )
  }
}

const Search = styled.div`
  ${mixins.flexRow};

  justify-content: center;
  align-items: center;
  height: 150px;
  background-color: ${colors.primary};

  ${mixins.mobile`
    height: 120px;
  `}
`

const FieldContainer = styled.div`
  ${mixins.flexRow};
  ${mixins.pageWidth};

  flex: 1;
  height: 56px;
  padding: 0 70px;

  ${mixins.mobile`
    padding: 0 30px;
  `}
`

const focused = css`
  border-width: 2px;
  transform: perspective(200px) translateZ(1px);
`

const Field = styled.div`
  ${mixins.flexRow};
  ${mixins.font.sizes.h5};

  flex: 1;
  border: solid ${colors.white};
  border-width: 1px;
  border-radius: 10px;
  overflow: hidden;
  transform: perspective(200px) translateZ(0px);
  color: ${colors.white};
  transition: border-width 0.15s, transform 0.15s;

  ${({ isFocused }) => isFocused && focused}
`

const Input = styled.input`
  flex: 1;
  padding: 0 25px;

  &:focus {
    cursor: text;
  }

  ::placeholder {
    color: ${colors.white};
    opacity: 0.5;
  }

  ${mixins.mobile`
    padding: 0 15px;
  `}
`

const Button = styled(Link)`
  ${mixins.flexRow};
  ${mixins.flexCenter};
  ${mixins.font.bold};

  width: 215px;
  color: ${colors.primary};
  background-color: ${colors.white};
  text-transform: uppercase;

  ${mixins.mobile`
    width: 90px;
  `}
`
