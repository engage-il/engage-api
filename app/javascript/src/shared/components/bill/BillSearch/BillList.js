// @flow
import * as React from 'react'
import { createPaginationContainer, graphql } from 'react-relay'
import type { RelayPaginationProps } from 'react-relay'
import { withRouter } from 'react-router-dom'
import type { ContextRouter } from 'react-router-dom'
import styled from 'react-emotion'
import { BillCell } from './BillCell'
import { BILL_SPACING, BILL_SPACING_MOBILE } from './constants'
import { TranslateAndFade, Button } from '@/components'
import { session } from '@/storage'
import { colors, mixins } from '@/styles'
import type { Viewer } from '@/types'

type Props = {
  viewer: Viewer,
  isAnimated: boolean,
  pageSize?: number
} & ContextRouter & RelayPaginationProps

type State = {
  disablesAnimation: boolean
}

function formatCount ({ bills }: Viewer) {
  return `Found ${bills.count} result${bills.count === 1 ? '' : 's'}.`
}

let BillList = class BillList extends React.Component<Props, State> {
  state = {
    disablesAnimation: this.isFromPop
  }

  // accessors
  get isFromPop () {
    return this.props.history.action === 'POP'
  }

  // events
  didClickLoadMore = () => {
    const { relay, pageSize } = this.props
    if (!relay.hasMore() || relay.isLoading()) {
      return
    }

    pageSize && relay.loadMore(pageSize, (error: ?Error) => {
      if (error) {
        console.error(`error loading next page: ${error.toString()}`)
      }
    })
  }

  // lifecycle
  componentDidMount () {
    if (this.isFromPop) {
      this.setState({ disablesAnimation: false })
    }
  }

  componentWillUnmount () {
    const { viewer } = this.props
    viewer && session.set('last-search-count', `${viewer.bills.edges.length}`)
  }

  render () {
    const {
      viewer,
      isAnimated,
      pageSize,
      relay
    } = this.props

    const {
      disablesAnimation
    } = this.state

    return (
      <Bills>
        <h5>{formatCount(viewer)}</h5>
        <TranslateAndFade
          component={List}
          disable={!isAnimated || disablesAnimation}
        >
          {viewer.bills.edges.map(({ node }) => (
            <BillCell key={node.id} bill={node} />
          ))}
        </TranslateAndFade>
        {pageSize && relay.hasMore() && (
          <ActionButton
            isSecondary
            onClick={this.didClickLoadMore}
            children='Load More'
          />
        )}
      </Bills>
    )
  }
}

BillList = withRouter(createPaginationContainer(BillList, graphql`
  fragment BillList_viewer on Viewer {
    bills(
      params: $params,
      first: $count,
      after: $cursor
    ) @connection(key: "BillList_bills") {
      count
      pageInfo {
        hasNextPage
        endCursor
      }
      edges {
        node {
          id
          ...BillCell_bill
        }
      }
    }
  }
`, {
  direction: 'forward',
  getConnectionFromProps ({ viewer }) {
    return viewer && viewer.bills
  },
  getVariables (props, { count, cursor }, previous) {
    return { ...previous, count, cursor }
  },
  query: graphql`
    query BillListQuery(
      $params: BillsSearchParams!,
      $count: Int!,
      $cursor: String!
    ) {
      viewer {
        ...BillList_viewer
      }
    }
  `
}))

const Bills = styled.div`
  ${mixins.flexColumn};

  > h5 {
    margin-bottom: ${BILL_SPACING}px;

    ${mixins.mobile`
      margin-bottom: ${BILL_SPACING_MOBILE}px;
    `}
  }
`

const List = styled.div`
  > * {
    border-bottom: 1px solid ${colors.gray4};
    margin-bottom: ${BILL_SPACING}px;
    padding-bottom: ${BILL_SPACING}px;

    ${mixins.mobile`
      margin-bottom: ${BILL_SPACING_MOBILE}px;
      padding-bottom: ${BILL_SPACING_MOBILE}px;
    `}
  }
`

const ActionButton = styled(Button)`
  align-self: center;
  margin-bottom: 70px;

  ${mixins.mobile`
    margin-bottom: 50px;
  `}
`

export { BillList }
