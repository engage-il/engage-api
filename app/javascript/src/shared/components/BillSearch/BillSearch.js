// @flow
import * as React from 'react'
import { createRefetchContainer, graphql } from 'react-relay'
import type { RelayRefetchProp } from 'react-relay'
import styled from 'react-emotion'
import { throttle } from 'lodash'
import { SearchField } from './SearchField'
import { BillList } from './BillList'
import { LoadingIndicator } from '../LoadingIndicator'
import type { Viewer } from 'shared/types'
import { mixins } from 'shared/styles'
import { addDays, endOfDay, startOfDay } from 'shared/date'

type Props = {
  viewer: ?Viewer,
  relay: RelayRefetchProp
}

type State = {
  query: string,
  disableAnimations: boolean
}

export const initialVariables = {
  query: '',
  count: 20,
  startDate: startOfDay(new Date()),
  endDate: endOfDay(addDays(new Date(), 6))
}

let BillSearch = class BillSearch extends React.Component<*, Props, State> {
  state = {
    query: initialVariables.query,
    disableAnimations: false
  }

  // actions
  filterBillsForQuery = throttle((query: string) => {
    const { relay } = this.props

    this.setState({ disableAnimations: true })
    relay.refetch({ query }, null, (error: ?Error) => {
      if (error) {
        console.error(`error updaing query: ${error.toString()}`)
      }

      // completion comes back before render
      global.requestAnimationFrame(() => {
        this.setState({ disableAnimations: false })
      })
    })
  }, 300)

  // events
  searchFieldDidChange = (query: string) => {
    this.setState({ query })
    this.filterBillsForQuery(query)
  }

  // lifecycle
  render () {
    const { viewer } = this.props
    const { query, disableAnimations } = this.state

    return (
      <Section>
        <SearchField
          value={query}
          onChange={this.searchFieldDidChange}
        />
        <Bills>
          <LoadingIndicator
            isLoading={!viewer}
          />
          {viewer && (
            <BillList
              viewer={viewer}
              animated={!disableAnimations}
            />
          )}
        </Bills>
      </Section>
    )
  }
}

BillSearch = createRefetchContainer(BillSearch,
  graphql`
    fragment BillSearch_viewer on Viewer {
      bills(
        first: $count, after: $cursor,
        query: $query, from: $startDate, to: $endDate
      ) {
        edges { node { id } }
      }
      ...BillList_viewer
    }
  `,
  graphql`
    query BillSearchQuery(
      $count: Int!, $cursor: String!,
      $query: String!, $startDate: Time!, $endDate: Time!
    ) {
      viewer {
        ...BillSearch_viewer
      }
    }
  `
)

const Section = styled.div`
  ${mixins.flexColumn};
`

const Bills = styled.div`
  position: relative;
`

export { BillSearch }
