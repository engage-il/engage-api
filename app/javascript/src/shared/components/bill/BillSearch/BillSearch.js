// @flow
import * as React from 'react'
import { createRefetchContainer, graphql } from 'react-relay'
import type { RelayRefetchProps } from 'react-relay'
import styled from 'react-emotion'
import { throttle } from 'lodash'
import { SearchField } from './SearchField'
import { SearchFilters } from './SearchFilters'
import { BillList } from './BillList'
import { BILL_SPACING, BILL_SPACING_MOBILE } from './constants'
import { Loading } from '@/components'
import { mixins } from '@/styles'
import type { Viewer, SearchParams } from '@/types'

type Props = {
  viewer: ?Viewer,
  pageSize?: number,
  params?: SearchParams,
  onChange?: (SearchParams) => void,
} & RelayRefetchProps

type State = {
  params: SearchParams,
  disablesAnimation: boolean
}

let BillSearch = class BillSearch extends React.Component<Props, State> {
  state = {
    params: {
      query: '',
      ...(this.props.params || {})
    },
    disablesAnimation: false
  }

  // actions
  filterBills = throttle((params: SearchParams) => {
    const { onChange, relay } = this.props

    // callback the change handler
    onChange && onChange(params)

    // fetch the bills with updated params
    this.setState({ disablesAnimation: true })

    // $FlowFixMe: variables failiing to match object
    relay.refetch({ params }, null, (error: ?Error) => {
      if (error) {
        console.error(`error updaing query: ${error.toString()}`)
      }

      // HACK: completion comes back before render
      global.requestAnimationFrame(() => {
        this.setState({ disablesAnimation: false })
      })
    })
  }, 300)

  // events
  didChangeParams = (params) => {
    this.setState({ params })
    this.filterBills(params)
  }

  // lifecycle
  render () {
    const {
      viewer,
      pageSize
    } = this.props

    const {
      params,
      disablesAnimation
    } = this.state

    return (
      <Search>
        <SearchField
          params={params}
          onChange={this.didChangeParams}
        />
        <Body>
          <SearchFilters
            params={params}
            onChange={this.didChangeParams}
          />
          <Bills>
            <Loading
              isLoading={!viewer}
            />
            {viewer && (
              <BillList
                viewer={viewer}
                pageSize={pageSize}
                isAnimated={!disablesAnimation}
              />
            )}
          </Bills>
        </Body>
      </Search>
    )
  }
}

BillSearch = createRefetchContainer(BillSearch,
  graphql`
    fragment BillSearch_viewer on Viewer {
      bills(
        params: $params,
        first: $count,
        after: $cursor
      ) @connection(key: "BillSearch_bills") {
        edges {
          node {
            id
          }
        }
      }
      ...BillList_viewer
    }
  `,
  graphql`
    query BillSearchQuery(
      $params: BillsSearchParams!,
      $count: Int!,
      $cursor: String!
    ) {
      viewer {
        ...BillSearch_viewer
      }
    }
  `
)

const Search = styled.div`
  ${mixins.flexColumn};
`

const Body = styled.div`
  ${mixins.pageContent};
  align-self: center;
`

const Bills = styled.div`
  position: relative;
  padding-top: ${BILL_SPACING}px;

  ${mixins.mobile`
    padding-top: ${BILL_SPACING_MOBILE}px;
  `}
`

export { BillSearch }
