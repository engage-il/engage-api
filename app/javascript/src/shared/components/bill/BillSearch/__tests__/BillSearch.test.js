/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { defaultsDeep } from 'lodash'
import { BillSearch } from '../BillSearch'

const { anything } = expect

// mocks
jest.mock('lodash', () => ({
  throttle: (fn) => fn
}))

// subject
let subject

const defaults = {
  viewer: 'test-viewer',
  pageSize: 'test-size',
  onChange: jest.fn(),
  relay: {
    refetch: jest.fn()
  }
}

function loadSubject (props = {}) {
  subject = shallow(<BillSearch {...defaultsDeep(props, defaults)} />)
}

// specs
afterEach(() => {
  subject = null
})

describe('#render', () => {
  it('shows the search view', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })

  it('shows the search view with default params', () => {
    loadSubject({
      params: {
        query: 'test-query'
      }
    })

    expect(subject).toMatchSnapshot()
  })

  it('shows the loading indicator when loading', () => {
    loadSubject({
      viewer: null
    })

    expect(subject).toMatchSnapshot()
  })
})

describe('changing the params', () => {
  const params = { test: 'query' }

  it('stores the params', () => {
    loadSubject()
    subject.instance().didChangeParams(params)
    expect(subject).toHaveState('params', params)
  })

  it('starts the re-fetch', () => {
    loadSubject()
    subject.instance().didChangeParams(params)
    expect(subject).toHaveState('disablesAnimation', true)
    expect(defaults.relay.refetch).toHaveBeenCalledWith({ params }, null, anything())
  })

  it('re-enables animations on completion', () => {
    jest.useFakeTimers()
    defaults.relay.refetch.mockImplementationOnce((_, _p, completion) => completion())

    loadSubject()
    subject.instance().didChangeParams(params)
    jest.runOnlyPendingTimers()

    expect(subject).toHaveState('disablesAnimation', false)
  })

  it('notifies the handler', () => {
    loadSubject()
    subject.instance().didChangeParams(params)
    expect(defaults.onChange).toHaveBeenCalledWith(params)
  })
})
