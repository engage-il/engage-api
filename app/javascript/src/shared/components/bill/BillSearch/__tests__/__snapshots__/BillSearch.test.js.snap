// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render shows the loading indicator when loading 1`] = `
<Search>
  <SearchField
    onChange={[Function]}
    params={
      Object {
        "query": "",
      }
    }
  />
  <Body>
    <SearchFilters
      onChange={[Function]}
      params={
        Object {
          "query": "",
        }
      }
    />
    <Bills>
      <Loading
        isLoading={true}
      />
    </Bills>
  </Body>
</Search>
`;

exports[`#render shows the search view 1`] = `
<Search>
  <SearchField
    onChange={[Function]}
    params={
      Object {
        "query": "",
      }
    }
  />
  <Body>
    <SearchFilters
      onChange={[Function]}
      params={
        Object {
          "query": "",
        }
      }
    />
    <Bills>
      <Loading
        isLoading={false}
      />
      <BillList
        isAnimated={true}
        pageSize="test-size"
        viewer="test-viewer"
      />
    </Bills>
  </Body>
</Search>
`;

exports[`#render shows the search view with default params 1`] = `
<Search>
  <SearchField
    onChange={[Function]}
    params={
      Object {
        "query": "test-query",
      }
    }
  />
  <Body>
    <SearchFilters
      onChange={[Function]}
      params={
        Object {
          "query": "test-query",
        }
      }
    />
    <Bills>
      <Loading
        isLoading={false}
      />
      <BillList
        isAnimated={true}
        pageSize="test-size"
        viewer="test-viewer"
      />
    </Bills>
  </Body>
</Search>
`;
